# **<font color="#ff5653">1_ETL - Extract, transform and load</font>**

En este Notebook se realizara el ETL de los archivos descargados de la pagina del ENACOM (Ente Nacional de Comunicaciones) de la Republica Argentina.

El **objetivo** es llegar a obtener dataframe limpios para realizarles un EDA mas profundo en un notebook aparte y tambien poder relacionar los datos en Power BI donde obtendremos los diferentes KPI solicitados.

- Orden de procesamiento de los archivos:

(*nombre del archivo.csv en nuestro Datasets = nombre de cada archivo que figura en ENACOM = archivo.csv realizado el ETL*)

1.   1_Internet_fijo_por_cada_100_ hogares_por_provincia.csv = Número de accesos al servicio de Internet fijo por cada 100 hogares por provincia = 1_internet_100hogares-por_provincia.csv

2.   11_Velocidades_internet_por_provincia_ultimo_trimestre.csv = Acceso a Internet Fijo por rangos de velocidad de bajada y provincia = 2_internet_rangovelocidades_velocidadm_provincia.csv

3. 15_Internet_fijo_por_tecnologia_por_localidad.csv = Accesos a Internet fijo por tecnología y Localidad = 3_internet_portecnologia_porlocalidad.csv

4. 4_Calendario_Fechas.csv

5. 6_Internet_fijo_por_tecnologia_por_provincia.csv = Acceso a Internet fijo por tecnología y provincia = 5_internet_portecnologia_porprovincia.csv



In [1]:
# Conexion entre Google Colab y Drive
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [2]:
# Libreria
import numpy as np
import pandas as pd
import csv

### **<font color="#ff5653">1. 1_Internet_fijo_por_cada_100_ hogares_por_provincia.csv</font>**

Para el procesamiento del dataframe se llevaran a cabo primero un EDA preliminar y despues el ETL.

- EDA (Exploratory Data Analysis - Análisis Exploratorio de Datos)

In [ ]:
# Cargamos el archivo
df_1=pd.read_csv("/content/drive/MyDrive/Datasets/1_Internet_fijo_por_cada_100_ hogares_por_provincia.csv")

In [ ]:
# Primeras dos filas del dataframe
df_1.head(2)

Año  Trimestre        Provincia Accesos por cada 100 hogares  Unnamed: 4  \
0  2022          4     Buenos Aires                        78,94         NaN   
1  2022          4  Capital Federal                       122,73         NaN   

   Unnamed: 5  Unnamed: 6  
0         NaN         NaN  
1         NaN         NaN

In [ ]:
# Cantidad de valores no nulos, tipos de datos, nombre de columnas.
df_1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 864 entries, 0 to 863
Data columns (total 7 columns):
 #   Column                        Non-Null Count  Dtype  
---  ------                        --------------  -----  
 0   Año                           864 non-null    int64  
 1   Trimestre                     864 non-null    int64  
 2   Provincia                     864 non-null    object 
 3   Accesos por cada 100 hogares  864 non-null    object 
 4   Unnamed: 4                    0 non-null      float64
 5   Unnamed: 5                    0 non-null      float64
 6   Unnamed: 6                    0 non-null      float64
dtypes: float64(3), int64(2), object(2)
memory usage: 47.4+ KB


In [ ]:
# Cantidad de valores nulos del dataframe
df_1.isna().sum()

Año                               0
Trimestre                         0
Provincia                         0
Accesos por cada 100 hogares      0
Unnamed: 4                      864
Unnamed: 5                      864
Unnamed: 6                      864
dtype: int64

In [ ]:
# Cantidad de filas y columnas del dataframe
df_1.shape

(864, 7)

In [ ]:
# Funcion: Valores unicos
def valores_unicos_columna(dataframe, columnas_especificas):

    for column in columnas_especificas:
        if column in dataframe.columns:
            valores_unicos = dataframe[column].unique()
            print(f"Valores únicos para la columna '{column}':", valores_unicos)

        else:
            print(f"La columna '{column}' no existe en el DataFrame.")

In [ ]:
# Llamamos a la funcion: Valores unicos.
c1 =["Año", "Trimestre", "Provincia"]                          # Solo para estas columnas vemos los valores unicos: Año, Trimestre y Provincia.
valores_unicos_columna(df_1,c1)

Valores únicos para la columna 'Año': [2022 2021 2020 2019 2018 2017 2016 2015 2014]
Valores únicos para la columna 'Trimestre': [4 3 2 1]
Valores únicos para la columna 'Provincia': ['Buenos Aires' 'Capital Federal' 'Catamarca' 'Chaco' 'Chubut' 'Córdoba'
 'Corrientes' 'Entre Ríos' 'Formosa' 'Jujuy' 'La Pampa' 'La Rioja'
 'Mendoza' 'Misiones' 'Neuquén' 'Río Negro' 'Salta' 'San Juan' 'San Luis'
 'Santa Cruz' 'Santa Fe' 'Santiago Del Estero' 'Tierra Del Fuego'
 'Tucumán']


In [ ]:
# Llamamos a la funcion: Valores unicos.
c1_1 =["Accesos por cada 100 hogares"]                          # Solo para esta columna vemos los valores unicos: Accesos por cada 100 hogares.
valores_unicos_columna(df_1,c1_1)

Valores únicos para la columna 'Accesos por cada 100 hogares': ['78,94' '122,73' '65,77' '43,62' '85,57' '89,03' '51,64' '64,01' '39,41'
 '58,46' '94,80' '79,86' '51,36' '52,85' '78,67' '68,42' '57,96' '50,88'
 '90,27' '43,53' '78,40' '48,05' '101,44' '61,04' '78,11' '122,28' '65,33'
 '43,86' '84,38' '88,31' '48,21' '63,67' '43,76' '58,91' '96,59' '80,66'
 '51,03' '51,82' '80,37' '67,98' '57,76' '50,80' '81,74' '43,19' '78,55'
 '47,98' '99,51' '61,14' '77,58' '121,31' '62,40' '43,79' '83,19' '87,48'
 '48,35' '63,90' '43,42' '57,32' '92,98' '80,14' '49,39' '51,00' '80,02'
 '67,85' '57,54' '49,46' '81,76' '43,15' '77,90' '47,70' '99,86' '59,31'
 '76,08' '111,80' '58,21' '44,06' '88,05' '85,76' '47,64' '63,70' '34,96'
 '58,06' '91,66' '78,77' '48,74' '49,04' '74,17' '67,04' '56,11' '43,55'
 '80,72' '43,34' '77,30' '47,17' '98,29' '57,39' '75,66' '111,27' '55,25'
 '44,05' '82,57' '84,51' '45,51' '63,16' '34,45' '56,92' '89,57' '77,92'
 '47,51' '49,06' '73,30' '65,69' '56,81' '39,59' '80,30

In [ ]:
# Resumen estadistico
df_1.describe()

Año   Trimestre  Unnamed: 4  Unnamed: 5  Unnamed: 6
count   864.000000  864.000000         0.0         0.0         0.0
mean   2018.000000    2.500000         NaN         NaN         NaN
std       2.583484    1.118682         NaN         NaN         NaN
min    2014.000000    1.000000         NaN         NaN         NaN
25%    2016.000000    1.750000         NaN         NaN         NaN
50%    2018.000000    2.500000         NaN         NaN         NaN
75%    2020.000000    3.250000         NaN         NaN         NaN
max    2022.000000    4.000000         NaN         NaN         NaN

- ETL (Extract, Transform and Load - Extraer, transformar, cargar)

En base a lo anterior se realizara: eliminacion de valores nulos, cambiamos los tipos de datos y agregamos una columna nueva.

In [ ]:
# Elimino columnas con valores nulos
df_1_1 = df_1.dropna(axis=1)


In [ ]:
# En la columna "Accesos por cada 100 hogares" cambiamos el tipo de dato a "float64"
df_1_1["Accesos por cada 100 hogares"] = df_1_1["Accesos por cada 100 hogares"].str.replace(',', '.').astype("float64")

In [ ]:
# Creamos una columna nueva "Año_trimestre" para relacionar los dataframe por esta columna.
df_1_1["Año_trimestre"] = (df_1_1["Trimestre"]*2 + (df_1_1["Trimestre"]-2)).astype(str) + "/" + "1"  + "/" + df_1_1["Año"].astype(str)

In [ ]:
# Cambiamos el tipo de dato de la columna "Año_Trimestre" a "datatime64"
df_1_1["Año_trimestre"] = pd.to_datetime(df_1_1["Año_trimestre"], errors="coerce")

In [ ]:
df_1_1.head(2)

Año  Trimestre        Provincia  Accesos por cada 100 hogares  \
0  2022          4     Buenos Aires                         78.94   
1  2022          4  Capital Federal                        122.73   

  Año_trimestre  
0    2022-10-01  
1    2022-10-01

In [ ]:
df_1_1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 864 entries, 0 to 863
Data columns (total 5 columns):
 #   Column                        Non-Null Count  Dtype         
---  ------                        --------------  -----         
 0   Año                           864 non-null    int64         
 1   Trimestre                     864 non-null    int64         
 2   Provincia                     864 non-null    object        
 3   Accesos por cada 100 hogares  864 non-null    float64       
 4   Año_trimestre                 864 non-null    datetime64[ns]
dtypes: datetime64[ns](1), float64(1), int64(2), object(1)
memory usage: 33.9+ KB


In [ ]:
# Guardamos el dataframe en formato .csv
df_1_1.to_csv("1_internet_100hogares_por_provincia.csv", index=False)

### **<font color="#ff5653">2. 11_Velocidades_internet_por_provincia_ultimo_trimestre.csv</font>**

Para el procesamiento del dataframe se llevaran a cabo primero un EDA preliminar y despues el ETL.

- EDA (Exploratory Data Analysis - Análisis Exploratorio de Datos)

In [ ]:
# Cargamos el archivo
df_11=pd.read_csv("/content/drive/MyDrive/Datasets/11_Velocidades_internet_por_provincia_ultimo_trimestre.csv")

In [ ]:
# Primeras dos filas del dataframe
df_11.head(2)

Año  Trimestre        Provincia  HASTA 512 kbps  + 512 Kbps - 1 Mbps  \
0  2022          4     Buenos Aires          28.934               23.347   
1  2022          4  Capital Federal         516.000                5.096   

  + 1 Mbps - 6 Mbps  + 6 Mbps - 10 Mbps  + 10 Mbps - 20 Mbps  \
0           259.866             281.550              244.555   
1            31.320              64.653               36.336   

   + 20 Mbps - 30 Mbps  + 30 Mbps    OTROS      Total  
0               98.670  3.760.109   97.634  4.794.665  
1                8.116  1.299.117  106.772  1.551.926

In [ ]:
# Cantidad de valores no nulos, tipos de datos, nombre de columnas.
df_11.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 864 entries, 0 to 863
Data columns (total 12 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   Año                  864 non-null    int64  
 1   Trimestre            864 non-null    int64  
 2   Provincia            864 non-null    object 
 3   HASTA 512 kbps       864 non-null    float64
 4   + 512 Kbps - 1 Mbps  864 non-null    float64
 5   + 1 Mbps - 6 Mbps    864 non-null    object 
 6   + 6 Mbps - 10 Mbps   864 non-null    float64
 7   + 10 Mbps - 20 Mbps  864 non-null    float64
 8   + 20 Mbps - 30 Mbps  864 non-null    float64
 9   + 30 Mbps            864 non-null    object 
 10  OTROS                858 non-null    float64
 11  Total                864 non-null    object 
dtypes: float64(6), int64(2), object(4)
memory usage: 81.1+ KB


In [ ]:
# Cantidad de valores nulos del dataframe
df_11.isna().sum()

Año                    0
Trimestre              0
Provincia              0
HASTA 512 kbps         0
+ 512 Kbps - 1 Mbps    0
+ 1 Mbps - 6 Mbps      0
+ 6 Mbps - 10 Mbps     0
+ 10 Mbps - 20 Mbps    0
+ 20 Mbps - 30 Mbps    0
+ 30 Mbps              0
OTROS                  6
Total                  0
dtype: int64

In [ ]:
# Cantidad de filas y columnas del dataframe
df_11.shape

(864, 12)

In [ ]:
# Resumen estadistico
df_11.describe()

Año   Trimestre  HASTA 512 kbps  + 512 Kbps - 1 Mbps  \
count   864.000000  864.000000      864.000000           864.000000   
mean   2018.000000    2.500000      126.069829           104.202722   
std       2.583484    1.118682      186.417712           211.789353   
min    2014.000000    1.000000        1.007000             0.000000   
25%    2016.000000    1.750000       11.000000             3.310750   
50%    2018.000000    2.500000       51.000000             8.762000   
75%    2020.000000    3.250000      133.250000            85.525500   
max    2022.000000    4.000000      998.000000           999.000000   

       + 6 Mbps - 10 Mbps  + 10 Mbps - 20 Mbps  + 20 Mbps - 30 Mbps  \
count          864.000000           864.000000           864.000000   
mean            70.717505            78.652155            99.993059   
std            138.831868           168.570068           207.349280   
min              0.000000             0.000000             0.000000   
25%              5.410250             5.000000             2.091000   
50%             20.208000            15.159000            11.042500   
75%             61.934500            51.959500            64.286250   
max            917.000000           978.000000           997.000000   

            OTROS  
count  858.000000  
mean    34.439517  
std    133.288022  
min     -1.945000  
25%      0.000000  
50%      0.000000  
75%      7.939750  
max    989.000000

- ETL (Extract, Transform and Load - Extraer, transformar, cargar)

En base a lo anterior se realizara: eliminar valores nulos, eliminamos los puntos de las columnas con datos numericos, cambiamos tipos de datos y creamos una columna "Año_trimestre".

In [ ]:
# Eliminamos valores nulos
df_11_1 = df_11.dropna()

In [ ]:
# Cambiamos el tipo de datos a string.
df_11_1['HASTA 512 kbps'] = df_11_1['HASTA 512 kbps'].astype(str)
df_11_1['+ 512 Kbps - 1 Mbps'] = df_11_1['+ 512 Kbps - 1 Mbps'].astype(str)
df_11_1['+ 1 Mbps - 6 Mbps'] = df_11_1['+ 1 Mbps - 6 Mbps'].astype(str)
df_11_1['+ 6 Mbps - 10 Mbps'] = df_11_1['+ 6 Mbps - 10 Mbps'].astype(str)
df_11_1['+ 10 Mbps - 20 Mbps'] = df_11_1['+ 10 Mbps - 20 Mbps'].astype(str)
df_11_1['+ 20 Mbps - 30 Mbps'] = df_11_1['+ 20 Mbps - 30 Mbps'].astype(str)
df_11_1['+ 30 Mbps'] = df_11_1['+ 30 Mbps'].astype(str)
df_11_1['OTROS'] = df_11_1['OTROS'].astype(str)
df_11_1['Total'] = df_11_1['Total'].astype(str)

In [ ]:
# Eliminamos los "."
df_11_1['HASTA 512 kbps'] = df_11_1['HASTA 512 kbps'].str.replace('.', '')
df_11_1['+ 512 Kbps - 1 Mbps'] = df_11_1['+ 512 Kbps - 1 Mbps'].str.replace('.', '')
df_11_1['+ 1 Mbps - 6 Mbps'] = df_11_1['+ 1 Mbps - 6 Mbps'].str.replace('.', '')
df_11_1['+ 6 Mbps - 10 Mbps'] = df_11_1['+ 6 Mbps - 10 Mbps'].str.replace('.', '')
df_11_1['+ 10 Mbps - 20 Mbps'] = df_11_1['+ 10 Mbps - 20 Mbps'].str.replace('.', '')
df_11_1['+ 20 Mbps - 30 Mbps'] = df_11_1['+ 20 Mbps - 30 Mbps'].str.replace('.', '')
df_11_1['+ 30 Mbps'] = df_11_1['+ 30 Mbps'].str.replace('.', '')
df_11_1['OTROS'] = df_11_1['OTROS'].str.replace('.', '')
df_11_1['Total'] = df_11_1['Total'].str.replace('.', '')

In [ ]:
# Cambiamos el tipo de datos a enteros.
df_11_1['HASTA 512 kbps'] = df_11_1['HASTA 512 kbps'].astype(int)
df_11_1['+ 512 Kbps - 1 Mbps'] = df_11_1['+ 512 Kbps - 1 Mbps'].astype(int)
df_11_1['+ 1 Mbps - 6 Mbps'] = df_11_1['+ 1 Mbps - 6 Mbps'].astype(int)
df_11_1['+ 6 Mbps - 10 Mbps'] = df_11_1['+ 6 Mbps - 10 Mbps'].astype(int)
df_11_1['+ 10 Mbps - 20 Mbps'] = df_11_1['+ 10 Mbps - 20 Mbps'].astype(int)
df_11_1['+ 20 Mbps - 30 Mbps'] = df_11_1['+ 20 Mbps - 30 Mbps'].astype(int)
df_11_1['+ 30 Mbps'] = df_11_1['+ 30 Mbps'].astype(int)
df_11_1['OTROS'] = df_11_1['OTROS'].astype(int)
df_11_1['Total'] = df_11_1['Total'].astype(int)

In [ ]:
# Creamos una columna nueva "Año_trimestre" para relacionar los dataframe por esta columna.
df_11_1["Año_trimestre"] = (df_11_1["Trimestre"]*2 + (df_11_1["Trimestre"]-2)).astype(str) + "/" + "1"  + "/" + df_11_1["Año"].astype(str)

In [ ]:
# Cambiamos el tipo de dato de la columna "Año_Trimestre" a "datatime64"
df_11_1["Año_trimestre"] = pd.to_datetime(df_11_1["Año_trimestre"], errors="coerce")

In [ ]:
# Confirmamos los cambios
df_11_1.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 858 entries, 0 to 863
Data columns (total 13 columns):
 #   Column               Non-Null Count  Dtype         
---  ------               --------------  -----         
 0   Año                  858 non-null    int64         
 1   Trimestre            858 non-null    int64         
 2   Provincia            858 non-null    object        
 3   HASTA 512 kbps       858 non-null    int64         
 4   + 512 Kbps - 1 Mbps  858 non-null    int64         
 5   + 1 Mbps - 6 Mbps    858 non-null    int64         
 6   + 6 Mbps - 10 Mbps   858 non-null    int64         
 7   + 10 Mbps - 20 Mbps  858 non-null    int64         
 8   + 20 Mbps - 30 Mbps  858 non-null    int64         
 9   + 30 Mbps            858 non-null    int64         
 10  OTROS                858 non-null    int64         
 11  Total                858 non-null    int64         
 12  Año_trimestre        858 non-null    datetime64[ns]
dtypes: datetime64[ns](1), int64(11), ob

In [ ]:
# Confirmamos los cambios
df_11_1.head(2)

Año  Trimestre        Provincia  HASTA 512 kbps  + 512 Kbps - 1 Mbps  \
0  2022          4     Buenos Aires           28934                23347   
1  2022          4  Capital Federal            5160                 5096   

   + 1 Mbps - 6 Mbps  + 6 Mbps - 10 Mbps  + 10 Mbps - 20 Mbps  \
0             259866               28155               244555   
1              31320               64653                36336   

   + 20 Mbps - 30 Mbps  + 30 Mbps   OTROS    Total Año_trimestre  
0                 9867    3760109   97634  4794665    2022-10-01  
1                 8116    1299117  106772  1551926    2022-10-01

La columna "Total" no poseen la suma correctas por lo que hay que agregar una nueva columna llamada "Total y OTROS" que sera la suma de las columnas que representan las velocidades. Y vamos a crear otra columna llamada "Total_accesos" donde estan comprendidas todas las columnas velocidades execpto la columna "OTROS", asi poder usar los datos de la columna "Total_accesos" y poder sacar la velocidad promedio.

- Velocidad Promedio = (0,512*AR1 + 0,7512*AR2 + 3,5*AR3 + 8*AR4 + 15*AR5 + 25*AR6 + 30*AR7)/Total de accesos reportados

Donde AR1 es el total de accesos reportados para el primer rango (Hasta 512 kbps); AR2 es el total de accesos reportados para el segundo rango (Entre 512 kbps y 1 Mbps) y así sucesivamente.

Desde el cuarto trimestre de 2017, VP se calcula como el promedio de las velocidades unitarias en Mbps ponderadas por el número de accesos reportados para cada velocidad.

Rangos de velocidad:

1.	Hasta 512 kbps
2.	Entre 512 kbps y 1 Mbps
3.	Entre 1 Mbps y 6 Mbps
4.	Entre 6 Mbps y 10 Mbps
5.	Entre 10 Mbps y 20 Mbps
6.	Entre 20 Mbps y 30 Mbps
7.	Más de 30 Mbps

Fuente: Ente Nacional de Comunicaciones (ENACOM), Argentina.

In [ ]:
# Creamos una columna nueva llamada "Total y OTROS" y sumamos las columnas con las velocidades.
df_11_1["Total y OTROS"] = df_11_1[["HASTA 512 kbps", "+ 512 Kbps - 1 Mbps", "+ 1 Mbps - 6 Mbps", "+ 6 Mbps - 10 Mbps", "+ 10 Mbps - 20 Mbps", "+ 20 Mbps - 30 Mbps", "+ 30 Mbps", "OTROS"]].sum(axis=1)

In [ ]:
# Creamos una columna nueva llamada "Total_accesos" y sumamos las columnas con las velocidades.
df_11_1["Total_accesos"] = df_11_1[["HASTA 512 kbps", "+ 512 Kbps - 1 Mbps", "+ 1 Mbps - 6 Mbps", "+ 6 Mbps - 10 Mbps", "+ 10 Mbps - 20 Mbps", "+ 20 Mbps - 30 Mbps", "+ 30 Mbps"]].sum(axis=1)

In [ ]:
df_11_1

Año  Trimestre            Provincia  HASTA 512 kbps  \
0    2022          4         Buenos Aires           28934   
1    2022          4      Capital Federal            5160   
2    2022          4            Catamarca             710   
3    2022          4                Chaco            4610   
4    2022          4               Chubut            1090   
..    ...        ...                  ...             ...   
859  2014          1           Santa Cruz            1610   
860  2014          1             Santa Fe            8456   
861  2014          1  Santiago Del Estero            1234   
862  2014          1     Tierra Del Fuego             120   
863  2014          1              Tucumán              60   

     + 512 Kbps - 1 Mbps  + 1 Mbps - 6 Mbps  + 6 Mbps - 10 Mbps  \
0                  23347             259866               28155   
1                   5096              31320               64653   
2                   3150               2716                5028   
3                   7710              16067               17846   
4                   1614              46457               31007   
..                   ...                ...                 ...   
859                 1625              24972                  10   
860               124468             345225               20328   
861                10531              22817                2422   
862                 6070              30902                  60   
863                34672              83210               11779   

     + 10 Mbps - 20 Mbps  + 20 Mbps - 30 Mbps  + 30 Mbps   OTROS    Total  \
0                 244555                 9867    3760109   97634  4794665   
1                  36336                 8116    1299117  106772  1551926   
2                   4638                 3122      52739     228    70909   
3                   6704                 3039      94866    3909   143663   
4                  34728                 1577      19048   20026   168759   
..                   ...                  ...        ...     ...      ...   
859                   10                    0          0       0    26760   
860                 6845                  230        668       0   506013   
861                 1090                    0          0       0    37113   
862                    0                    0          0       0    31527   
863                 3620                   30          0       0   130032   

    Año_trimestre  Total y OTROS  Total_accesos  
0      2022-10-01        4452467        4354833  
1      2022-10-01        1556570        1449798  
2      2022-10-01          72331          72103  
3      2022-10-01         154751         150842  
4      2022-10-01         155547         135521  
..            ...            ...            ...  
859    2014-01-01          28227          28227  
860    2014-01-01         506220         506220  
861    2014-01-01          38094          38094  
862    2014-01-01          37152          37152  
863    2014-01-01         133371         133371  

[858 rows x 15 columns]

- Al comprar las columnas "Total" y "Total y OTROS", queda claro las diferencias que hay y el error que tenia el archivo original.

- Con la columna "Total_accesos" vamos a utilizarla para sacar la velocidad media de conexion utilizando la formula proporcionada por el ENACOM.

In [ ]:
# Eliminamos la columna "Total general": porque los datos no reflejan la sumatorias de las tecnologias.
df_11_1.drop(columns=["Total"], inplace=True)

In [ ]:
df_11_2 = df_11_1

In [ ]:
df_11_2

Año  Trimestre            Provincia  HASTA 512 kbps  \
0    2022          4         Buenos Aires           28934   
1    2022          4      Capital Federal            5160   
2    2022          4            Catamarca             710   
3    2022          4                Chaco            4610   
4    2022          4               Chubut            1090   
..    ...        ...                  ...             ...   
859  2014          1           Santa Cruz            1610   
860  2014          1             Santa Fe            8456   
861  2014          1  Santiago Del Estero            1234   
862  2014          1     Tierra Del Fuego             120   
863  2014          1              Tucumán              60   

     + 512 Kbps - 1 Mbps  + 1 Mbps - 6 Mbps  + 6 Mbps - 10 Mbps  \
0                  23347             259866               28155   
1                   5096              31320               64653   
2                   3150               2716                5028   
3                   7710              16067               17846   
4                   1614              46457               31007   
..                   ...                ...                 ...   
859                 1625              24972                  10   
860               124468             345225               20328   
861                10531              22817                2422   
862                 6070              30902                  60   
863                34672              83210               11779   

     + 10 Mbps - 20 Mbps  + 20 Mbps - 30 Mbps  + 30 Mbps   OTROS  \
0                 244555                 9867    3760109   97634   
1                  36336                 8116    1299117  106772   
2                   4638                 3122      52739     228   
3                   6704                 3039      94866    3909   
4                  34728                 1577      19048   20026   
..                   ...                  ...        ...     ...   
859                   10                    0          0       0   
860                 6845                  230        668       0   
861                 1090                    0          0       0   
862                    0                    0          0       0   
863                 3620                   30          0       0   

    Año_trimestre  Total y OTROS  Total_accesos  
0      2022-10-01        4452467        4354833  
1      2022-10-01        1556570        1449798  
2      2022-10-01          72331          72103  
3      2022-10-01         154751         150842  
4      2022-10-01         155547         135521  
..            ...            ...            ...  
859    2014-01-01          28227          28227  
860    2014-01-01         506220         506220  
861    2014-01-01          38094          38094  
862    2014-01-01          37152          37152  
863    2014-01-01         133371         133371  

[858 rows x 14 columns]

In [ ]:
# Velocidad media de conexion
df_11_2["velocidad_media_conexion(Mbps)"] = (0.512 * df_11_2["HASTA 512 kbps"] + 0.7512 * df_11_2["+ 512 Kbps - 1 Mbps"] + 3.5 * df_11_2["+ 1 Mbps - 6 Mbps"] + 8 * df_11_2["+ 6 Mbps - 10 Mbps"] + 15 * df_11_2["+ 10 Mbps - 20 Mbps"] + 25 * df_11_2["+ 20 Mbps - 30 Mbps"] + 30 * df_11_2["+ 30 Mbps"]) / df_11_2["Total_accesos"]

In [ ]:
df_11_3 = df_11_2

In [ ]:
# Redondeamos los valores de "velocidad_media_conexion(Mbps)".
df_11_3["velocidad_media_conexion(Mbps)"] = df_11_3["velocidad_media_conexion(Mbps)"].astype(int)

In [ ]:
df_11_3

Año  Trimestre            Provincia  HASTA 512 kbps  \
0    2022          4         Buenos Aires           28934   
1    2022          4      Capital Federal            5160   
2    2022          4            Catamarca             710   
3    2022          4                Chaco            4610   
4    2022          4               Chubut            1090   
..    ...        ...                  ...             ...   
859  2014          1           Santa Cruz            1610   
860  2014          1             Santa Fe            8456   
861  2014          1  Santiago Del Estero            1234   
862  2014          1     Tierra Del Fuego             120   
863  2014          1              Tucumán              60   

     + 512 Kbps - 1 Mbps  + 1 Mbps - 6 Mbps  + 6 Mbps - 10 Mbps  \
0                  23347             259866               28155   
1                   5096              31320               64653   
2                   3150               2716                5028   
3                   7710              16067               17846   
4                   1614              46457               31007   
..                   ...                ...                 ...   
859                 1625              24972                  10   
860               124468             345225               20328   
861                10531              22817                2422   
862                 6070              30902                  60   
863                34672              83210               11779   

     + 10 Mbps - 20 Mbps  + 20 Mbps - 30 Mbps  + 30 Mbps   OTROS  \
0                 244555                 9867    3760109   97634   
1                  36336                 8116    1299117  106772   
2                   4638                 3122      52739     228   
3                   6704                 3039      94866    3909   
4                  34728                 1577      19048   20026   
..                   ...                  ...        ...     ...   
859                   10                    0          0       0   
860                 6845                  230        668       0   
861                 1090                    0          0       0   
862                    0                    0          0       0   
863                 3620                   30          0       0   

    Año_trimestre  Total y OTROS  Total_accesos  \
0      2022-10-01        4452467        4354833   
1      2022-10-01        1556570        1449798   
2      2022-10-01          72331          72103   
3      2022-10-01         154751         150842   
4      2022-10-01         155547         135521   
..            ...            ...            ...   
859    2014-01-01          28227          28227   
860    2014-01-01         506220         506220   
861    2014-01-01          38094          38094   
862    2014-01-01          37152          37152   
863    2014-01-01         133371         133371   

     velocidad_media_conexion(Mbps)  
0                                27  
1                                27  
2                                24  
3                                21  
4                                11  
..                              ...  
859                               3  
860                               3  
861                               3  
862                               3  
863                               3  

[858 rows x 15 columns]

In [ ]:
# Guardamos el dataframe en formato .csv
df_11_3.to_csv("2_internet_rangovelocidad_velocidadm_provincia.csv", index=False)

### **<font color="#ff5653">3. 15_Internet_fijo_por_tecnologia_por_localidad.csv</font>**
Para el procesamiento del dataframe se llevaran a cabo primero un EDA preliminar y despues el ETL.

In [ ]:
# Cargamos el archivo
df_15 = pd.read_csv("/content/drive/MyDrive/Datasets/15_Internet_fijo_por_tecnologia_por_localidad.csv")

In [ ]:
# P rimeras dos filas del dataframe
df_15.head(2)

Provincia     Partido   Localidad Link Indec ADSL CABLEMODEM DIAL UP  \
0  BUENOS AIRES  25 de Mayo  25 de Mayo    6854100  874       4,54     - 0   
1  BUENOS AIRES  25 de Mayo   Del Valle    6854020  181        - 0     - 0   

  FIBRA OPTICA OTROS SATELITAL WIMAX WIRELESS Total general  Unnamed: 13  
0            3   - 0       852   - 0      706          6,98          NaN  
1          - 0    10         1   - 0      - 0           192          NaN

In [ ]:
# Cantidad de valores no nulos, tipos de datos, nombre de columnas.
df_15.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3137 entries, 0 to 3136
Data columns (total 14 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Provincia      3137 non-null   object 
 1   Partido        3137 non-null   object 
 2   Localidad      3137 non-null   object 
 3   Link Indec     3137 non-null   object 
 4   ADSL           3137 non-null   object 
 5   CABLEMODEM     3137 non-null   object 
 6   DIAL UP        3137 non-null   object 
 7   FIBRA OPTICA   3137 non-null   object 
 8   OTROS          3137 non-null   object 
 9   SATELITAL      3137 non-null   object 
 10  WIMAX          3137 non-null   object 
 11  WIRELESS       3137 non-null   object 
 12  Total general  3137 non-null   object 
 13  Unnamed: 13    0 non-null      float64
dtypes: float64(1), object(13)
memory usage: 343.2+ KB


In [ ]:
# Cantidad de valores nulos del dataframe
df_15.isna().sum()

Provincia           0
Partido             0
Localidad           0
Link Indec          0
ADSL                0
CABLEMODEM          0
DIAL UP             0
FIBRA OPTICA        0
OTROS               0
SATELITAL           0
WIMAX               0
WIRELESS            0
Total general       0
Unnamed: 13      3137
dtype: int64

In [ ]:
# Cantidad de filas y columnas del dataframe
df_15.shape

(3137, 14)

In [ ]:
# Resumen estadistico
df_15.describe()

Unnamed: 13
count          0.0
mean           NaN
std            NaN
min            NaN
25%            NaN
50%            NaN
75%            NaN
max            NaN

- ETL (Extract, Transform and Load - Extraer, transformar, cargar)

En base a lo anterior se realizara: eliminacion de valores nulos, cambiamos "-0", "," , los tipos de datos y agregamos una columna nueva.

In [ ]:
# Eliminamos la columna con valores nulos
df_15_1 = df_15.dropna(axis=1)

In [ ]:
# Cambiamos el "- 0" por " 0"

df_15_1['ADSL'] = df_15_1['ADSL'].str.replace('- 0', '0')
df_15_1['CABLEMODEM'] = df_15_1['CABLEMODEM'].str.replace('- 0', '0')
df_15_1['DIAL UP'] = df_15_1['DIAL UP'].str.replace('- 0', '0')
df_15_1['FIBRA OPTICA'] = df_15_1['FIBRA OPTICA'].str.replace('- 0', '0')
df_15_1['OTROS'] = df_15_1['OTROS'].str.replace('- 0', '0')
df_15_1['SATELITAL'] = df_15_1['SATELITAL'].str.replace('- 0', '0')
df_15_1['WIMAX'] = df_15_1['WIMAX'].str.replace('- 0', '0')
df_15_1['WIRELESS'] = df_15_1['WIRELESS'].str.replace('- 0', '0')
df_15_1['Total general'] = df_15_1['Total general'].str.replace('- 0', '0')
df_15_1['Link Indec'] = df_15_1['Link Indec'].str.replace('Sin Datos', '0')        # Reemplazamos 'Sin Datos' por "0". Asi nos permite cambiar el tipo de dato a entero.

In [ ]:
# Eliminamos las ","
df_15_1['ADSL'] = df_15_1['ADSL'].str.replace(',', '')
df_15_1['CABLEMODEM'] = df_15_1['CABLEMODEM'].str.replace(',', '')
df_15_1['DIAL UP'] = df_15_1['DIAL UP'].str.replace(',', '')
df_15_1['FIBRA OPTICA'] = df_15_1['FIBRA OPTICA'].str.replace(',', '')
df_15_1['OTROS'] = df_15_1['OTROS'].str.replace(',', '')
df_15_1['SATELITAL'] = df_15_1['SATELITAL'].str.replace(',', '')
df_15_1['WIMAX'] = df_15_1['WIMAX'].str.replace(',', '')
df_15_1['WIRELESS'] = df_15_1['WIRELESS'].str.replace(',', '')
df_15_1['Total general'] = df_15_1['Total general'].str.replace(',', '')
df_15_1['Link Indec'] = df_15_1['Link Indec'].str.replace(',', '')

In [ ]:
# Cambiamos el tipo de dato a entero.
df_15_1['ADSL'] = df_15_1['ADSL'].astype(int)
df_15_1['CABLEMODEM'] = df_15_1['CABLEMODEM'].astype(int)
df_15_1['DIAL UP'] = df_15_1['DIAL UP'].astype(int)
df_15_1['FIBRA OPTICA'] = df_15_1['FIBRA OPTICA'].astype(int)
df_15_1['OTROS'] = df_15_1['OTROS'].astype(int)
df_15_1['SATELITAL'] = df_15_1['SATELITAL'].astype(int)
df_15_1['WIMAX'] = df_15_1['WIMAX'].astype(int)
df_15_1['WIRELESS'] = df_15_1['WIRELESS'].astype(int)
df_15_1['Total general'] = df_15_1['Total general'].astype(int)
df_15_1['Link Indec'] = df_15_1['Link Indec'].astype(int)

In [ ]:
# Creamos una columna nueva llamada "Total General" y sumamos las columnas con las tecnologias.
df_15_1["Total General"] = df_15_1[["ADSL", "CABLEMODEM", "DIAL UP", "FIBRA OPTICA", "OTROS", "SATELITAL", "WIMAX", "WIRELESS"]].sum(axis=1)

In [ ]:
df_15_1.head(2)

Provincia     Partido   Localidad  Link Indec  ADSL  CABLEMODEM  \
0  BUENOS AIRES  25 de Mayo  25 de Mayo     6854100   874         454   
1  BUENOS AIRES  25 de Mayo   Del Valle     6854020   181           0   

   DIAL UP  FIBRA OPTICA  OTROS  SATELITAL  WIMAX  WIRELESS  Total general  \
0        0             3      0        852      0       706            698   
1        0             0     10          1      0         0            192   

   Total General  
0           2889  
1            192

In [ ]:
# Eliminamos la columna "Total general": porque los datos no reflejan la sumatorias de las tecnologias.
df_15_1.drop(columns=["Total general"], inplace=True)

In [ ]:
df_15_1

Provincia      Partido                           Localidad  \
0     BUENOS AIRES   25 de Mayo                          25 de Mayo   
1     BUENOS AIRES   25 de Mayo                           Del Valle   
2     BUENOS AIRES   25 de Mayo                   Gobernador Ugarte   
3     BUENOS AIRES   25 de Mayo              Norberto de la Riestra   
4     BUENOS AIRES   25 de Mayo                    Lucas Monteverde   
...            ...          ...                                 ...   
3132       TUCUMAN   Tafí Viejo  Villa Mariano Moreno - El Colmenar   
3133       TUCUMAN      Trancas                San Pedro de Colalao   
3134       TUCUMAN      Trancas                   Villa  de Trancas   
3135       TUCUMAN  Yerba Buena                       Villa Carmela   
3136       TUCUMAN  Yerba Buena            Yerba Buena - Marcos Paz   

      Link Indec  ADSL  CABLEMODEM  DIAL UP  FIBRA OPTICA  OTROS  SATELITAL  \
0        6854100   874         454        0             3      0        852   
1        6854020   181           0        0             0     10          1   
2        6854040     0           0        0             0      0          0   
3        6854060     0         782        0             6    167          0   
4        6854050     0           0        0             0      0          0   
...          ...   ...         ...      ...           ...    ...        ...   
3132    90105100     0          31        0           373      0          0   
3133    90112020     0           0        0             0      0          0   
3134    90112030   126           0        0             0     33          0   
3135    90119020    15           0        0             0      1          0   
3136    90119030   129         209        0           154     33          0   

      WIMAX  WIRELESS  Total General  
0         0       706           2889  
1         0         0            192  
2         0       181            181  
3         0       327           1282  
4         0         6              6  
...     ...       ...            ...  
3132      0         0            404  
3133      0       313            313  
3134      0       101            260  
3135      0        97            113  
3136      0         0            525  

[3137 rows x 13 columns]

In [ ]:
df_15_1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3137 entries, 0 to 3136
Data columns (total 13 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   Provincia      3137 non-null   object
 1   Partido        3137 non-null   object
 2   Localidad      3137 non-null   object
 3   Link Indec     3137 non-null   int64 
 4   ADSL           3137 non-null   int64 
 5   CABLEMODEM     3137 non-null   int64 
 6   DIAL UP        3137 non-null   int64 
 7   FIBRA OPTICA   3137 non-null   int64 
 8   OTROS          3137 non-null   int64 
 9   SATELITAL      3137 non-null   int64 
 10  WIMAX          3137 non-null   int64 
 11  WIRELESS       3137 non-null   int64 
 12  Total General  3137 non-null   int64 
dtypes: int64(10), object(3)
memory usage: 318.7+ KB


In [ ]:
# Guardamos el dataframe en formato .csv
df_15_1.to_csv("3_internet_portecnologia_porlocalidad.csv", index=False)

### **<font color="#ff5653">4. 4_Calendario_Fechas</font>**

Se creara un calendario de fechas por treimestres desde el año 2014 al 2023.
El objetivo de esta tabla es ser el nexo entre todos los dataframe.

In [ ]:
# Generar la secuencia de fechas trimestrales
fechas = pd.date_range(start="01/01/2014", end="31/12/2022")

# Crear el DataFrame vacío
calendario = pd.DataFrame()

# Asignar las fechas a una columna llamada Año_Trimestre
calendario_fechas = calendario.assign(**{"Año_Trimestre": fechas})

<ipython-input-61-9b235c0eae12>:2: UserWarning: Parsing dates in DD/MM/YYYY format when dayfirst=False (the default) was specified. This may lead to inconsistently parsed dates! Specify a format to ensure consistent parsing.
  fechas = pd.date_range(start="01/01/2014", end="31/12/2022")


In [ ]:
calendario_fechas

Año_Trimestre
0       2014-01-01
1       2014-01-02
2       2014-01-03
3       2014-01-04
4       2014-01-05
...            ...
3282    2022-12-27
3283    2022-12-28
3284    2022-12-29
3285    2022-12-30
3286    2022-12-31

[3287 rows x 1 columns]

In [ ]:
calendario_fechas.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3287 entries, 0 to 3286
Data columns (total 1 columns):
 #   Column         Non-Null Count  Dtype         
---  ------         --------------  -----         
 0   Año_Trimestre  3287 non-null   datetime64[ns]
dtypes: datetime64[ns](1)
memory usage: 25.8 KB


In [ ]:
# Crear dos columnas nuevas "Año" y "Trimestre"
calendario_fechas["Año"] = calendario_fechas["Año_Trimestre"].dt.year                    # Toma el dato de año de la columna "Año_Trimestre"
calendario_fechas["Trimestre"] = calendario_fechas["Año_Trimestre"].dt.quarter           # Toma el dato del cuatrimestre de la columna "Año_Trimestre"

In [ ]:
calendario_fechas

Año_Trimestre   Año  Trimestre
0       2014-01-01  2014          1
1       2014-01-02  2014          1
2       2014-01-03  2014          1
3       2014-01-04  2014          1
4       2014-01-05  2014          1
...            ...   ...        ...
3282    2022-12-27  2022          4
3283    2022-12-28  2022          4
3284    2022-12-29  2022          4
3285    2022-12-30  2022          4
3286    2022-12-31  2022          4

[3287 rows x 3 columns]

In [ ]:
# Guardamos el dataframe en formato .csv
calendario_fechas.to_csv("4_Calendario_Fechas.csv", index=False)

### **<font color="#ff5653">5. 6_Internet_fijo_por_tecnologia_por_provincia.csv</font>**

Para el procesamiento del dataframe se llevaran a cabo primero un EDA preliminar y despues el ETL.

- EDA (Exploratory Data Analysis - Análisis Exploratorio de Datos)

In [3]:
df_6=pd.read_csv("/content/drive/MyDrive/Datasets/6_Internet_fijo_por_tecnologia_por_provincia.csv")

In [ ]:
# Primeras dos filas del dataframe
df_6.head(2)

Año Trimestre        Provincia   ADSL  Cablemodem  Fibra óptica  Wireless  \
0  2022         4     Buenos Aires  326.0         3.0           1.0     130.0   
1  2022         4  Capital Federal  139.0         1.0         139.0       5.0   

   Otros  Total  
0   72.0    5.0  
1   32.0    2.0

In [ ]:
# Cantidad de valores no nulos, tipos de datos, nombre de columnas.
df_6.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 865 entries, 0 to 864
Data columns (total 9 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   Año           865 non-null    object 
 1   Trimestre     865 non-null    object 
 2   Provincia     864 non-null    object 
 3   ADSL          864 non-null    float64
 4   Cablemodem    864 non-null    float64
 5   Fibra óptica  864 non-null    float64
 6   Wireless      864 non-null    float64
 7   Otros         864 non-null    float64
 8   Total         864 non-null    float64
dtypes: float64(6), object(3)
memory usage: 60.9+ KB


In [ ]:
# Cantidad de valores nulos del dataframe
df_6.isna().sum()

Año             0
Trimestre       0
Provincia       1
ADSL            1
Cablemodem      1
Fibra óptica    1
Wireless        1
Otros           1
Total           1
dtype: int64

In [ ]:
# Cantidad de filas y columnas del dataframe
df_6.shape

(865, 9)

In [ ]:
# Funcion: Valores unicos
def valores_unicos_columna(dataframe, columnas_especificas):

    for column in columnas_especificas:
        if column in dataframe.columns:
            valores_unicos = dataframe[column].unique()
            print(f"Valores únicos para la columna '{column}':", valores_unicos)

        else:
            print(f"La columna '{column}' no existe en el DataFrame.")

In [ ]:
# Llamamos a la funcion: Valores unicos.
c5 =["Año", "Trimestre", "Provincia"]                          # Solo para estas columnas vemos los valores unicos: Año, Trimestre y Provincia.
valores_unicos_columna(df_6,c5)

Valores únicos para la columna 'Año': ['2022' '2021' '2020' '2019' '2019 *' '2018' '2017' '2016' '2015' '2014'
 '*']
Valores únicos para la columna 'Trimestre': ['4' '3' '2' '1' '3 *' '2 *' '1 *'
 'Los datos provinciales no coinciden a nivel nacional, ya que se rincorporó información que no contien apertuta a nivel geográfico.']
Valores únicos para la columna 'Provincia': ['Buenos Aires' 'Capital Federal' 'Catamarca' 'Chaco' 'Chubut' 'Córdoba'
 'Corrientes' 'Entre Ríos' 'Formosa' 'Jujuy' 'La Pampa' 'La Rioja'
 'Mendoza' 'Misiones' 'Neuquén' 'Río Negro' 'Salta' 'San Juan' 'San Luis'
 'Santa Cruz' 'Santa Fe' 'Santiago Del Estero' 'Tierra Del Fuego'
 'Tucumán' nan]


In [ ]:
# Llamamos a la funcion: Valores unicos
c5_1 = ["ADSL","Cablemodem","Fibra óptica","Wireless","Otros","Total"]
valores_unicos_columna(df_6,c5_1)

Valores únicos para la columna 'ADSL': [326. 139.   9.  25.  45. 152.  34.  56.  14.  18.   8.   6.  38.  33.
  29.  37.  39.  10. 164.  13.  69. 340. 141.  27.  60.  19.   7.  36.
  40. 184.  15.  78. 351. 145.  12.  30. 179.  64.  20.  43. 198.  17.
  86. 392. 102.  32.  44. 196.  67.  16.  21.  49.  28.  11. 213.  93.
 426. 111.  46. 212.  70.  22.  42.  47. 233.  98. 586. 143. 229.  74.
  24.  91.  48. 246. 104. 603. 149. 240.  92.  51. 257. 110. 630. 160.
  26. 260.  84.  55. 277. 147. 636. 163. 267.  50.  57. 287. 654. 170.
 276.  88.  58. 295. 150. 661. 283.  89. 300. 673. 286.  90. 302. 720.
 197.  23. 297.  53.  97.  63.  41. 312.   1. 308.  95.  54.  66.  52.
 319. 151. 310. 315.  68. 325. 153. 328. 323. 101. 159. 329. 336. 334.
  59. 338. 103.  77. 339. 161. 342. 345. 105. 350. 349. 107.  80. 571.
 108.  81. 157. 583. 154. 615. 330.   2. 429. 109. 327. 435. 337.  79.
 448. 348. 113. 335. 451. 454. 144. 455. 354. 114. 352. 453. 333. 140.
 344. 112.  76. 138. 456.  31. 137. 45

In [ ]:
# Resumen estadistico
df_6.describe()

ADSL  Cablemodem  Fibra óptica    Wireless       Otros  \
count  864.000000  864.000000    864.000000  864.000000  864.000000   
mean    86.363426  103.291667    165.432870   57.777778  102.712963   
std    116.431490  188.849172    266.807807  156.015038  207.418981   
min      1.000000    0.000000      0.000000    0.000000    1.000000   
25%     21.000000    9.000000      9.000000    4.000000    3.000000   
50%     46.000000   35.500000     25.000000   10.000000   15.000000   
75%     81.000000   72.250000    170.750000   26.000000   47.000000   
max    720.000000  999.000000    998.000000  952.000000  987.000000   

            Total  
count  864.000000  
mean   137.636574  
std    180.798813  
min      1.000000  
25%     39.750000  
50%     91.000000  
75%    143.250000  
max    987.000000

- ETL (Extract, Transform and Load - Extraer, transformar, cargar)

Eliminamos valores nulos, eliminamos asteriscos, cambiamos tipo de datos y creamos una nueva columna.

In [4]:
# Eliminamos valores nulos
df_6_1 = df_6.dropna()

In [ ]:
# Cambio tipo de dato.
df_6_1['Año'] = df_6_1['Año'].astype(str)
df_6_1['Trimestre'] = df_6_1['Trimestre'].astype(str)
df_6_1['Provincia'] = df_6_1['Provincia'].astype(str)
df_6_1['ADSL'] = df_6_1['ADSL'].astype(str)
df_6_1['Cablemodem'] = df_6_1['Cablemodem'].astype(str)
df_6_1['Fibra óptica'] = df_6_1['Fibra óptica'].astype(str)
df_6_1['Wireless'] = df_6_1['Wireless'].astype(str)
df_6_1['Otros'] = df_6_1['Otros'].astype(str)
df_6_1['Total'] = df_6_1['Total'].astype(str)

In [ ]:
# Elimino los Asteriscos.
df_6_1['Año'] = df_6_1['Año'].str.replace('*', '')
df_6_1['Trimestre'] = df_6_1['Trimestre'].str.replace('*', '')
df_6_1['Provincia'] = df_6_1['Provincia'].str.replace('*', '')
df_6_1['ADSL'] = df_6_1['ADSL'].str.replace('*', '')
df_6_1['Cablemodem'] = df_6_1['Cablemodem'].str.replace('*', '')
df_6_1['Fibra óptica'] = df_6_1['Fibra óptica'].str.replace('*', '')
df_6_1['Wireless'] = df_6_1['Wireless'].str.replace('*', '')
df_6_1['Otros'] = df_6_1['Otros'].str.replace('*', '')
df_6_1['Total'] = df_6_1['Total'].str.replace('*', '')

In [ ]:
# Redondeo los valores
df_6_1['ADSL'] = df_6_1['ADSL'].str.split('.').str[0]
df_6_1['Cablemodem'] = df_6_1['Cablemodem'].str.split('.').str[0]
df_6_1['Fibra óptica'] = df_6_1['Fibra óptica'].str.split('.').str[0]
df_6_1['Wireless'] = df_6_1['Wireless'].str.split('.').str[0]
df_6_1['Otros'] = df_6_1['Otros'].str.split('.').str[0]
df_6_1['Total'] = df_6_1['Total'].str.split('.').str[0]

In [ ]:
# Cambio tipo de dato
df_6_1['Año'] = df_6_1['Año'].astype(int)
df_6_1['Trimestre'] = df_6_1['Trimestre'].astype(int)
#df_6['Provincia'] = df_6['Provincia'].astype(str)

In [ ]:
# Cambio tipo de dato
df_6_1['ADSL'] = df_6_1['ADSL'].astype(int)
df_6_1['Cablemodem'] = df_6_1['Cablemodem'].astype(int)
df_6_1['Fibra óptica'] = df_6_1['Fibra óptica'].astype(int)
df_6_1['Wireless'] = df_6_1['Wireless'].astype(int)
df_6_1['Otros'] = df_6_1['Otros'].astype(int)
df_6_1['Total'] = df_6_1['Total'].astype(int)

In [ ]:
# Creamos una columna nueva "Año_trimestre" para relacionar los dataframe por esta columna.
df_6_1["Año_trimestre"] = (df_6_1["Trimestre"]*2 + (df_6_1["Trimestre"]-2)).astype(str) + "/" + "1"  + "/" + df_6_1["Año"].astype(str)

In [ ]:
# Cambiamos el tipo de dato de la columna "Año_Trimestre"
df_6_1["Año_trimestre"] = pd.to_datetime(df_6_1["Año_trimestre"], errors="coerce")

In [ ]:
# Llamamos a la funcion: Valores unicos.
c5_1 =["Año", "Trimestre", "Provincia"]                          # Solo para estas columnas vemos los valores unicos: Año, Trimestre y Provincia.
valores_unicos_columna(df_6_1,c5_1)

In [ ]:
# Creamos una nueva columna "Total_real"
df_6_1["Total_real"] = df_6_1[["ADSL", "Cablemodem", "Fibra óptica", "Wireless", "Otros"]].sum(axis=1)

In [15]:
df_6_1

Año  Trimestre            Provincia  ADSL  Cablemodem  Fibra óptica  \
0    2022          4         Buenos Aires   326           3             1   
1    2022          4      Capital Federal   139           1           139   
2    2022          4            Catamarca     9          11            47   
3    2022          4                Chaco    25          61            47   
4    2022          4               Chubut    45          74            10   
..    ...        ...                  ...   ...         ...           ...   
859  2014          1           Santa Cruz    19         100           381   
860  2014          1             Santa Fe   323         174             3   
861  2014          1  Santiago Del Estero    33           4            19   
862  2014          1     Tierra Del Fuego    22           3           648   
863  2014          1              Tucumán   130          83           121   

     Wireless  Otros  Total Año_trimestre  Total_real  
0         130     72      5    2022-10-01         532  
1           5     32      2    2022-10-01         316  
2           1      2     71    2022-10-01          70  
3           8      3    144    2022-10-01         144  
4          31      9    169    2022-10-01         169  
..        ...    ...    ...           ...         ...  
859         4      3     27    2014-01-01         507  
860         6    644    507    2014-01-01        1150  
861       915     34     37    2014-01-01        1005  
862         1      3     28    2014-01-01         677  
863        13     98    130    2014-01-01         445  

[864 rows x 11 columns]

In [ ]:
# Eliminamos la columna "Total_real": porque los datos no reflejan la sumatorias de las tecnologias.
df_6_1.drop(columns=["Total"], inplace=True)

In [18]:
df_6_1

Año  Trimestre            Provincia  ADSL  Cablemodem  Fibra óptica  \
0    2022          4         Buenos Aires   326           3             1   
1    2022          4      Capital Federal   139           1           139   
2    2022          4            Catamarca     9          11            47   
3    2022          4                Chaco    25          61            47   
4    2022          4               Chubut    45          74            10   
..    ...        ...                  ...   ...         ...           ...   
859  2014          1           Santa Cruz    19         100           381   
860  2014          1             Santa Fe   323         174             3   
861  2014          1  Santiago Del Estero    33           4            19   
862  2014          1     Tierra Del Fuego    22           3           648   
863  2014          1              Tucumán   130          83           121   

     Wireless  Otros Año_trimestre  Total_real  
0         130     72    2022-10-01         532  
1           5     32    2022-10-01         316  
2           1      2    2022-10-01          70  
3           8      3    2022-10-01         144  
4          31      9    2022-10-01         169  
..        ...    ...           ...         ...  
859         4      3    2014-01-01         507  
860         6    644    2014-01-01        1150  
861       915     34    2014-01-01        1005  
862         1      3    2014-01-01         677  
863        13     98    2014-01-01         445  

[864 rows x 10 columns]

In [19]:
# Guardamos el dataframe en formato .csv
df_6_1.to_csv("5_internet_portecnologia_porprovincia.csv", index=False)